clang-tools-extra/clang-tidy/bugprone/ReturnConstRefFromParameterCheck.cpp

```diff
@@ -0,0 +1,34 @@
+//===--- ReturnConstRefFromParameterCheck.cpp - clang-tidy ----------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "ReturnConstRefFromParameterCheck.h"
+#include "../utils/Matchers.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+#include "clang/ASTMatchers/ASTMatchers.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::bugprone {
+
+void ReturnConstRefFromParameterCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(
+      returnStmt(hasReturnValue(declRefExpr(to(parmVarDecl(hasType(
+                     hasCanonicalType(matchers::isReferenceToConst())))))))
+          .bind("ret"),
+      this);
```

Comment:
Does this need to check that the function's return type is a constant reference as well? For example, is
```c++
const S fn(const S &a) {
    return a;
}
```
without the reference return type okay? The documentation of this check says so:

> Detects return statements that return a constant reference parameter **as constant reference**.

As I understand the AUTOSAR rule, it's about pointers or references to automatic variables returned from a function. So

```c++
int *f(int i) { return &i; }
```

and

```c++
int *g() {
    int i = 1;
    return &i;
}
```

would be violations of the rule too.

I know that we are not allowed to cite the rules literally, but only checking constant reference parameters might be too restricted. Or are there other checks implementing the other cases already? If so, should they be extended instead of having a new separate check?

Reply:
clang will create `CXXConstructExpr` for object and `ImplicitCastExpr` for builtin type. just like
```
`-FunctionDecl 0x13d8cf548 <line:3:1, col:36> col:9 fn 'const S (const S &)'
  |-ParmVarDecl 0x13d8cf450 <col:12, col:21> col:21 used a 'const S &'
  `-CompoundStmt 0x13d8ee1a0 <col:24, col:36>
    `-ReturnStmt 0x13d8ee190 <col:26, col:33>
      `-CXXConstructExpr 0x13d8ee160 <col:33> 'const S':'const S' 'void (const S &) noexcept'
        `-DeclRefExpr 0x13d8cf640 <col:33> 'const S':'const S' lvalue ParmVar 0x13d8cf450 'a' 'const S &'
```
and 
```
`-FunctionDecl 0x1410d9ac0 <line:3:1, col:40> col:11 fn 'const int (const int &)'
  |-ParmVarDecl 0x1410d99f0 <col:14, col:25> col:25 used a 'const int &'
  `-CompoundStmt 0x1410d9c00 <col:28, col:40>
    `-ReturnStmt 0x1410d9bf0 <col:30, col:37>
      `-ImplicitCastExpr 0x1410d9bd8 <col:37> 'int' <LValueToRValue>
        `-DeclRefExpr 0x1410d9bb8 <col:37> 'const int' lvalue ParmVar 0x1410d99f0 'a' 'const int &'
```

Reply:
stack variable leak can be detected by clang diagnose directly, no need to use clang-tidy.

Reply:
I see. Thank you for clarification!